natss/pkg/reconciler/dispatcher/natsschannel_test.go

```diff
@@ -0,0 +1,202 @@
+*/
+
+package controller
+
+import (
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	fakekubeclientset "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	clientgotesting "k8s.io/client-go/testing"
+	"knative.dev/pkg/controller"
+	logtesting "knative.dev/pkg/logging/testing"
+	. "knative.dev/pkg/reconciler/testing"
+
+	clientset "knative.dev/eventing-contrib/natss/pkg/client/clientset/versioned"
+	informers "knative.dev/eventing-contrib/natss/pkg/client/informers/externalversions"
+	dispatchertesting "knative.dev/eventing-contrib/natss/pkg/dispatcher/testing"
+	"knative.dev/eventing-contrib/natss/pkg/reconciler"
+	reconciletesting "knative.dev/eventing-contrib/natss/pkg/reconciler/testing"
+)
+
+const (
+	testNS = "test-namespace"
+	ncName = "test-nc"
+)
+
+func TestAllCases(t *testing.T) {
+	ncKey := testNS + "/" + ncName
+
+	table := TableTest{
+		{
+			Name: "bad workqueue key",
+			// Make sure Reconcile handles bad keys.
+			Key: "too/many/parts",
+		},
+		{
+			Name: "key not found",
+			// Make sure Reconcile handles good keys that don't exist.
+			Key: "foo/not-found",
+		},
+		{
+			Name: "reconcile ok: channel ready",
+			Key:  ncKey,
+			Objects: []runtime.Object{
+				reconciletesting.NewNatssChannel(ncName, testNS,
+					reconciletesting.WithReady,
+				),
+			},
+			WantPatches: []clientgotesting.PatchActionImpl{
+				makeFinalizerPatch(testNS, ncName),
+			},
+			WantErr: false,
+		},
+		{
+			Name: "reconcile error: channel not ready",
+			Key:  ncKey,
+			Objects: []runtime.Object{
+				reconciletesting.NewNatssChannel(ncName, testNS,
+					reconciletesting.WithNotReady("ups", ""),
+				),
+			},
+			WantErr: true,
+		},
+		{
+			Name: "remove finalizer even though channel is not ready",
+			Key:  ncKey,
+			Objects: []runtime.Object{
+				reconciletesting.NewNatssChannel(ncName, testNS,
+					// the finalizer should get removed
+					reconciletesting.WithNatssChannelFinalizer,
+					// make sure that finalizer can get removed even if channel is not ready
+					reconciletesting.WithNotReady("ups", ""),
+					reconciletesting.WithNatssChannelDeleted),
+			},
+			WantErr: false,
+			WantUpdates: []clientgotesting.UpdateActionImpl{
+				{
+					Object: reconciletesting.NewNatssChannel(ncName, testNS,
+						// finalizer is removed
+						// make sure that finalizer can get removed even if channel is not ready
+						reconciletesting.WithNotReady("ups", ""),
+						reconciletesting.WithNatssChannelDeleted),
+				},
+			},
+		},
+	}
+	defer logtesting.ClearAll()
+
+	table.Test(t, reconciletesting.MakeFactory(func(listers *reconciletesting.Listers, opt reconciler.Options) controller.Reconciler {
+		natssDispatcher := dispatchertesting.NewDispatcherDoNothing(t)
+
+		r := Reconciler{
+			Base:                 reconciler.NewBase(opt, controllerAgentName),
+			natssDispatcher:      natssDispatcher,
+			natsschannelLister:   listers.GetNatssChannelLister(),
+			natsschannelInformer: nil,
```

Comment:
See comment above, I'd avoid having `nil`s (no pun intended, Nils) in the test Reconciler (no mocks).

Reply:
Will be removed since `natsschannelInformer` is not required for the struct

Reply:
Will be removed since natsschannelInformer is not required for the struct.